File: in_Java/0003 Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> hm = new HashMap<>();
        int begin = 0, end = 0, counter = 0, d = 0;
        while(end < s.length()){
            char c = s.charAt(end);
            hm.put(c, hm.getOrDefault(c,0)+1);
            if(hm.get(c) > 1) counter++;
            end++;
            
            while(counter != 0){
                char temp = s.charAt(begin);
                if(hm.get(temp) > 1) counter--;
                hm.put(temp, hm.get(temp)-1);
                begin++;
            }
            d = Math.max(d, end - begin);
        }
        return d;
    }
}

File: in_Python/0003 Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        N = len(s)
        i,j = 0,0
        res = 0
        d = {}
        for i in range(N):
            c = s[i]
            if c not in d or d[c] == -1:
                d[c] = i
                res = max(res, i - j + 1)
            else:
                res = max(res, i - j)
                j = d[c] + 1
                m = d[c]
                for k in d:
                    if d[k] <= m:
                        d[k] = -1     
                d[c] = i
        return res